docs/code/SchemaComprehension.md

```diff
@@ -0,0 +1,210 @@
+```
+It is important to note that the `PredictionEngine` actually *validates* that the 'pipeline' conforms to the input and output schema requirements when it is created.
+
+The same can be said about the `PredictorModel<InputType, OutputType>`. This is a somewhat more restricted version of `PredictionEngine` that is created by `LearningPipeline.Train`.
+
+Please note that **`PredictionEngine` and `PredictorModel` are not thread-safe**: they hold an internal cursor object, and therefore cannot be used in a re-entrant fashion.
+If you ever see the error message that says: `An attempt was made to keep iterating after the pipe has been reset`, it most likely means that ML.NET has detected a race condition on the `PredictionEngine`.
+
+## Type system mapping
+
+`IDataView` [type system](IDataViewTypeSystem.md) differs slightly from the C# type system, so a 1-1 mapping between column types and C# types is not always feasible. 
+Below are the most notable examples of the differences:
+
+* `IDataView` vector columns may have a fixed (and known) size, C# arrays can not. You can use `[VectorType(N)]` attribute to an array field to specify that the column is a vector of fixed size N. This is often necessary: most ML components don't work with variable-size vectors, they require fixed-size ones.
+* `IDataView`'s **key types** don't have an underlying C# type either. To declare a key-type column, you need to make your field an `uint`, and decorate it with `[KeyType(Min=A, Count=B)]` to denote that the field is a key with the specified range of values.
+
+### Full list of type mappings
+The below table illustrates what C# types are mapped to what `IDataView` types:
+
+| `IDataView` type | C# type     |  C# type with extra conversion |
+| ---------------- | ----------- | ------------------------------ |
+| `I1`             | `DvInt1`    | `sbyte`, `sbyte?`              |
+| `I2`             | `DvInt2`    | `short`, `short?`              |
+| `I4`             | `DvInt4`    | `int`, `int?`                  |
+| `I8`             | `DvInt8`    | `long`, `long?`                |
+| `U1`             | `byte`      | `byte?`                        |
+| `U2`             | `ushort`    | `ushort?`                      |
+| `U4`             | `uint`      | `uint?`                        |
+| `U8`             | `ulong`     | `ulong?`                       |
+| `UG`             | `UInt128`   |                                |
+| `R4`             | `float`     | `float?`                       |
+| `R8`             | `double`    | `double?`                      |
+| `TX`             | `DvText`, `string` |                         |
+| `BL`             | `DvBool`           | `bool`, `bool?`         |
+| `TS`             | `DvTimeSpan`       |                         |
+| `DT`             | `DvDateTime`       |                         |
+| `DZ`             | `DvDateTimeZone`   |                         |
```

Comment:
They don't map to TimeSpan, DateTime and DataTimeZone?

Reply:
No

---
In reply to: [204836126](https://github.com/dotnet/machinelearning/pull/572#discussion_r204836126) [](ancestors = 204836126)